src/ConnectedMode.UnitTests/ServerSentEvents/SSESessionManagerTests.cs

```diff
@@ -0,0 +1,235 @@
+    {
+        var testScope = new TestScope();
+
+
+        if (switchToConnectedMode)
+        {
+            var projectKey = "myproj";
+            testScope.SetUpSSEFactoryToReturnNoOpSSESession(projectKey);
+            testScope.SwitchToConnectedMode(projectKey);
+        }
+        else
+        {
+            testScope.SwitchToStandaloneMode();
+        }
+
+        testScope.SSESessionFactoryMock.Verify(factory => factory.Create(It.IsAny<string>()),
+            Times.Exactly(switchToConnectedMode ? 1 : 0));
+    }
+
+    [TestMethod]
+    public void OnSolutionChanged_WhenChangesFromClosedToOpen_CreatesSessionAndLaunchesIt()
+    {
+        var testScope = new TestScope();
+        testScope.SwitchToStandaloneMode();
+        var projectKey = "myproj";
+        var sessionMock = testScope.SetUpSSEFactoryToReturnNoOpSSESession(projectKey);
+
+        testScope.SwitchToConnectedMode(projectKey);
+
+        testScope.SSESessionFactoryMock.Verify(factory => factory.Create(It.IsAny<string>()), Times.Once);
+        sessionMock.Verify(session => session.PumpAllAsync(), Times.Once);
+    }
+
+    [TestMethod]
+    public void OnSolutionChanged_WhenChangesFromOpenToClosed_CancelsSession()
+    {
+        var testScope = new TestScope();
+        var projectKey = "myproj";
+        var sessionMock = testScope.SetUpSSEFactoryToReturnNoOpSSESession(projectKey);
+        sessionMock.Setup(session => session.Dispose());
+        testScope.SwitchToConnectedMode(projectKey);
+
+        testScope.SwitchToStandaloneMode();
+
+        sessionMock.Verify(session => session.Dispose(), Times.Once);
+    }
+
+    [TestMethod]
+    public void OnSolutionChanged_WhenChangesFromOpenToOpen_CancelsSessionAndStartsNewOne()
+    {
+        var testScope = new TestScope();
+        var projectKey1 = "proj1";
+        var projectKey2 = "proj2";
+        var sessionMock1 = testScope.SetUpSSEFactoryToReturnNoOpSSESession(projectKey1);
+        sessionMock1.Setup(session => session.Dispose());
+        testScope.SwitchToConnectedMode(projectKey1);
+        var sessionMock2 = testScope.SetUpSSEFactoryToReturnNoOpSSESession(projectKey2);
+
+        testScope.SwitchToConnectedMode(projectKey2);
+
+        sessionMock1.Verify(session => session.Dispose(), Times.Once);
+        sessionMock2.Verify(session => session.Dispose(), Times.Never);
+    }
+
+    [TestMethod]
+    public void Dispose_WhenProjectOpened_CorrectAndIdempotent()
+    {
+        var testScope = new TestScope();
+        var projectKey = "myproj";
+        var sseSession = testScope.SetUpSSEFactoryToReturnNoOpSSESession(projectKey);
+        testScope.SetUpCorrectDisposeOrder(sseSession);
+        testScope.SwitchToConnectedMode(projectKey);
+
+        CallDisposeMultipleTimes(testScope);
+
+        CheckEverythingIsDisposedOnce(testScope);
+    }
+
+    [TestMethod]
+    public void Dispose_WhenProjectClosed_CorrectAndIdempotent()
```

Comment:
_WhenSessionDoesNotExist_... - I'm not sure what it's supposed to dispose and `CheckEverythingIsDisposedOnce` is really not clear

Reply:
I changed it